server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1645,39 +1648,96 @@ public void canMatch(CanMatchNodeRequest request, ActionListener<CanMatchNodeRes
      * won't match any documents on the current shard.
      */
     public CanMatchShardResponse canMatch(ShardSearchRequest request) throws IOException {
-        return canMatch(request, true);
+        ReaderContext readerContext = findReaderContext(request.readerId(), request);
+        IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
+        try (CanMatchContext canMatchContext = new CanMatchContext(request, indexService, readerContext, getKeepAlive(request))) {
+            return canMatch(canMatchContext, true);
+        }
+    }
+
+    static class CanMatchContext implements Releasable {
```

Comment:
most of the changes in SearchService have a simple goal: make the canMatch method unit testable. Given how difficult it is to recreate a SearchService instance (no wonder it requires spinning up a node like we do in SearchServiceTests), I went for making the method static and have some extensible way to provide what it needs as an argument.